File: clojure-solutions/src/clojure_solutions/day6.clj
(ns clojure-solutions.day6)

(defn- solve [n]
  (->> (slurp "./input/day6.txt")
       (partition n 1)
       (take-while #(not= (count (into #{} %)) (count %)))
       count
       (+ n)))

(defn day6 [part]
  (case part
    :one (solve 4)
    :two (solve 14)))

File: rust-solutions/src/day6.rs
use crate::util::*;
use std::collections::HashSet;

pub fn day6(p: Part) -> usize {
    match p {
        Part::One => solve(4),
        Part::Two => solve(14),
    }
}

fn solve(n: usize) -> usize {
    let inp: Vec<char> = std::fs::read_to_string("./input/day6.txt")
        .unwrap()
        .chars()
        .collect();
    n + inp
        .windows(n)
        .take_while(|win| {
            let mut hs = HashSet::new();
            win.iter().any(|x| !(hs.insert(x)))
        })
        .count()
}
